crates/bevy_pbr/src/ssr/mod.rs

```diff
@@ -0,0 +1,563 @@
+/// the depth and deferred prepasses required to enable them.
+#[derive(Bundle, Default)]
+pub struct ScreenSpaceReflectionsBundle {
+    /// The component that enables SSR.
+    pub settings: ScreenSpaceReflectionsSettings,
+    /// The depth prepass, needed for SSR.
+    pub depth_prepass: DepthPrepass,
+    /// The deferred prepass, needed for SSR.
+    pub deferred_prepass: DeferredPrepass,
+}
+
+/// Add this component to a camera to enable *screen-space reflections* (SSR).
+///
+/// Screen-space reflections currently require deferred rendering in order to
+/// appear. Therefore, you'll generally need to add a [`DepthPrepass`] and a
+/// [`DeferredPrepass`] to the camera as well.
+///
+/// SSR currently performs no roughness filtering for glossy reflections, so
+/// only very smooth surfaces will reflect objects in screen space. You can
+/// adjust the `perceptual_roughness_threshold` in order to tune the threshold
+/// below which screen-space reflections will be traced.
+///
+/// As with all screen-space techniques, SSR can only reflect objects on screen.
+/// When objects leave the camera, they will disappear from reflections.
+/// Alternatives that don't suffer from this problem include
+/// [`crate::environment_map::ReflectionProbeBundle`]s. The advantage of SSR is
+/// that it can reflect all objects, not just static ones.
+///
+/// SSR is an approximation technique and produces artifacts in some situations.
+/// Hand-tuning the settings in this component will likely be useful.
+///
+/// Screen-space reflections are presently unsupported on WebGL 2 because of a
+/// bug whereby Naga doesn't generate correct GLSL when sampling depth buffers,
+/// which is required for screen-space raymarching.
+#[derive(Clone, Copy, Component, Reflect)]
+#[reflect(Component, Default)]
+pub struct ScreenSpaceReflectionsSettings {
+    /// The maximum PBR roughness level that will enable screen space
+    /// reflections.
+    pub perceptual_roughness_threshold: f32,
+
+    /// When marching the depth buffer, we only have 2.5D information and don't
+    /// know how thick surfaces are. We shall assume that the depth buffer
+    /// fragments are cuboids with a constant thickness defined by this
+    /// parameter.
+    pub thickness: f32,
+
+    /// The number of steps to be taken at regular intervals to find an initial
+    /// intersection. Must not be zero.
+    ///
+    /// Higher values result in higher-quality reflections, because the
+    /// raymarching shader is less likely to miss objects. However, they take
+    /// more GPU time.
+    pub linear_steps: u32,
+
+    /// Exponent to be applied in the linear part of the march.
+    ///
+    /// A value of 1.0 will result in equidistant steps, and higher values will
+    /// compress the earlier steps, and expand the later ones. This might be
+    /// desirable in order to get more detail close to objects.
+    ///
+    /// For optimal performance, this should be a small unsigned integer, such
```

Comment:
this is not compile time expanded like the other comment suggests, is this optimization future work?

Reply:
I don't think our `shader_defs` allow for `#define`s with values, do they? So I don't think I have the option to do that.

Reply:
if only 1 or 2 is ever used, it could be a simple gate, although it does feel a bit ugly. im fine with merge as-is regardless, this should be noted somewhere though

Reply:
I'm not sure the decrease in maintainability is worth it. Do you want me to do performance tests?

Reply:
shader_defs support can have values from the cpu side. We also support this syntax inside a shader file `#define PI 3.1416` and you need to use it with `#{PI}`.

Reply:
not particularly, there's better things to do with your time :)

Reply:
Actually, in the original `raymarch.hlsl` this value isn't a compile time constant: https://gist.github.com/h3r2tic/9c8356bdaefbe80b1a22ae0aaee192db#file-raymarch-hlsl-L554

So I don't think it matters. I'll leave it as is unless there are objections.